rfd/0139-gogo-must-go.md

```diff
@@ -0,0 +1,279 @@
+  backend serialized in the wire format.
+
+##### Write custom storage objects for the backend
+
+Rather than use gRPC's wire format or serialized JSON format, we should consider writing
+a storage object format. This has a number of benefits:
+
+- The backend continues to be easy to read and modify.
+- We decouple our protobuf definitions from our storage representation, which makes changes
+  to the frontend less impactful.
+- This can be done without making any changes to the proto definitions.
+- It gives us more control over storage representation.
+- It makes migration easier, as we can easily maintain multiple versions of an object within
+  the Teleport codebase.
+
+The downsides:
+
+- It adds to the boilerplate necessary for an object.
+
+##### Write custom marshalers for objects
+
+In order to keep our user facing representations of the objects, we will need to write
+custom JSON and YAML marshalers for our objects. This may be time intensive but not overall
+a significant amount of work.
+
+#### Embedding types (removing `gogoproto.embed`)
+
+We heavily use `gogoproto.embed` for generating objects with a consistent structure or interface.
+We particularly use these with our `kind`, `version`, and `metadata` fields. This will be
+difficult to mitigate without some heavy lifting. As such, I recommend the custom storage object
+format listed above and the addition of `FromV*` functions in order to convert protobuf objects
+into this object. For example:
+
+```go
+  accessRequest := accessrequest.FromV1Proto(request.AccessRequest)
+  user := user.FromV6Proto(request.User)
+```
+
+#### Casting types (removing `gogoproto.casttype`)
+
+We additionally rely heavily on `gogoproto.casttype`, which allows us to avoid doing explicit
+casts or conversions within go code. We will have to take on this conversion work, unfortunately.
+
+#### Other gogoproto tags
+
+##### `gogoproto.nullable`
+
+We will need to remove the `gogoproto.nullable` tags, which overall shouldn't be a huge impact.
+We will need to rely on our `CheckAndSetDefaults` functions, which should generally handle this
+today.
+
+##### `gogoproto.stdtime`
+
+We will need to migrate to using `timestamppb` directly instead of relying on gogo's `stdtime` tag.
+
+### Proposed solution
+
+#### Create internal objects
+
+Currently, in `api/types` we define a large number of interfaces that are subsequently used to wrap
+our protobuf created messages in a nicer, more user friendly way, abstracting away much of the protobuf
+bits and pieces.
+
+We should break this relationship and create actual internal objects to replace interfaces. From here, we
+should use conversion functions to translate between protobuf messages and our internal objects.
+
+It's imperative that these objects are compatible with the current JSON/YAML representations in Teleport.
+That is, the objects that are in Teleport today should unmarshal properly into our new internal objects.
```

Comment:
Should we freeze the current resource types and require a version bump (to a "new world" version that uses protojson in the backend and a dedicated service for RPCs) for any future change, to avoid accidental mistakes slipping through, or do you think we can chug along for a while manually adding new fields to the conversion functions?

Reply:
That seems like a good idea. I'm less sold on the `protojson` purely for the reasons I described in https://github.com/gravitational/teleport/pull/28386#discussion_r1247659578.